staging/src/k8s.io/component-base/logs/options.go

```diff
@@ -75,6 +75,10 @@ func flagIsSet(name string) bool {
 	if f != nil {
 		return f.DefValue != f.Value.String()
 	}
+	f = flag.Lookup(strings.Replace(name, "-", "_", -1))
```

Comment:
can we use 

doing this transformation here seems fragile/wrong... does this mean the transformation in unsupportedLoggingFlags() is incorrect?

Reply:
It's reverse of one in `unsupportedLoggingFlags`.
One of the places we do this normalization https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/staging/src/k8s.io/component-base/cli/globalflag/globalflags.go#L52

Reply:
> It's reverse of one in unsupportedLoggingFlags.

right... special-casing this to undo a transformation done there seems fragile. It seems like the caller of this function should pass the correct flag name, rather than making this function transform and fallback...

Reply:
Yes, I didn't prioritize it thou as we are also working on unifying klog flag setup. https://github.com/kubernetes/kubernetes/issues/99265

Possibly we should take one step ahead, wrap klog by `component-base/logs` and forbid direct import of `klog` library. That way all the klog flags will be controlled in one place in k/k. So we avoid fragile copying. This is bigger effort thou

Reply:
I don't understand why we need to do the normalization at all, this seems to be a new breaking change in 1.21. Can we just leave the klogs flags alone and revert all these `_ -> -` changes? Personally I think we should have rejected https://github.com/kubernetes/kubernetes/pull/98524 as it is an `ACTION REQUIRED:` end user change for the (minimal) benefit of flag format consistency.

Reply:
#98524 didn't introduce any breaking changes. It's goal was to fix inconsistency between help message in --logging-format command and flags registered in components.

Reply:
Where are they getting normalized then? How did a help text fix break a flag? This is why I am confounded.

Reply:
They are normalized here https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/staging/src/k8s.io/component-base/cli/globalflag/globalflags.go#L45
Assuming that components calls globalflags.AddGlobalFlags instead of klog.Init

Reply:
As for why we need to both handle names with and without normalization, reason is that depending on how klog is initialized those flags will be registered under different names.

Reply:
@liggitt Is the proposed approach for a fix acceptable? If we would want to remove fragile string replace logic before code freeze we should start working on it early.

Reply:
I'm not seeing why this validation method needs to look up flags at all... can it do this and sidestep the whole issue?

```
func (o *Options) Validate() []error {
	errs := []error{}
	if _, err := o.Get(); err != nil {
		errs = append(errs, fmt.Errorf("unsupported log format: %s", o.LogFormat))
	} else if o.LogFormat != defaultLogFormat {
		errs = append(errs, fmt.Errorf("non-default logging format %q is not supported", o.LogFormat))
	}
	return errs
}
```

Reply:
It looks up flags because json format doesn't support all the flags that klog implements. When adding json format I was requested to protect users from trying to use klog specific features with json logging.

Reply:
> It looks up flags because json format doesn't support all the flags that klog implements. When adding json format I was requested to protect users from trying to use klog specific features with json logging.

I see, I misunderstood the flag/format interaction. Would something like this work instead:

```go
	if o.LogFormat != defaultLogFormat {
		allFlags := unsupportedLoggingFlags()
		for _, fname := range allFlags {
			f := flag.Lookup(fname)
			pf := pflag.Lookup(fname)
			if (f != nil && f.DefValue != f.Value.String()) || (pf != nil && pf.DefValue != pf.Value.String()) {
				errs = append(errs, fmt.Errorf("non-default logging format %q doesn't honor flag: %s", o.LogFormat, fname))
			}
		}
	}
```

Reply:
> It looks up flags because json format doesn't support all the flags that klog implements. When adding json format I was requested to protect users from trying to use klog specific features with json logging.

I see, I misunderstood the flag/format interaction. Yes, we have to ensure unsupported options are not set incorrectly and ignored.

> @liggitt Is the proposed approach for a fix acceptable? If we would want to remove fragile string replace logic before code freeze we should start working on it early.

I don't think it is acceptable. This is already really hard to follow, I don't think we should make it worse, making assumptions about the global pflag normalization functions and undoing changes we made in `unsupportedLoggingFlags`.

I think `unsupportedLoggingFlags` should probably be told how to normalize flags, and the two callers can pass in the normalization function for their respective flagsets, rather than hard-coding `_` and `-` replacements.

Reply:
@liggitt @serathius I have modified the code based on the review comments.